### ISSUE TITLE
Expose the OnError function

### ISSUE THREAD

**Issue by TimLariviere** [elmish/elmish#182]
When porting Fabulous to Elmish, I needed to access the function `onError` stored in `Program`.

Fabulous needs to report errors when dealing with view updates and the easiest way is to use the same pipeline as Elmish.

So the developer can just do
```fsharp
Program.mkSimple ...
|> Program.withConsoleTrace
|> FabulousProgram.run
```
and have both the output of Elmish and Fabulous.

A simple function would be sufficient to permit what I want to achieve:
```fsharp
module Program =
    /// For library authors only: return the onError function
    let onError (program: Program<'arg, 'model, 'msg, 'view>) =
        program.onError
```

Today, I need to use a dirty hack to access it:
```fsharp
type private ProgramAccessor<'model, 'msg>(program: Program<unit, 'model, 'msg, ViewElement>) =        
    member __.OnError =
        let mutable value = ignore
        Program.mapErrorHandler (fun onError -> value <- onError; onError) program |> ignore
        value
```

Would that be possible to add it to Elmish?

**Comment by MangelMaxime** [comment]
A similar issue was reported here https://github.com/elmish/elmish/issues/172

**Comment by TimLariviere** [comment]
I understand users will have to provide their error handlers in the end.
But as a library author, the answer of #172 is not good for me.

I need to make sure that the user set their error handler once for both Elmish and Fabulous.
It will always be the same output because it will happen in the same update loop (Fabulous handles the update of the view after Elmish called the update function).

But if I have to write my own logging mecanism, it will only end up as a duplicate of Elmish because Fabulous needs the exact same thing (I want to allow my users to choose where they trace their logs).

And if I reimplement this logging mecanism, I will also need to completely hide the Elmish.Program module.

Because otherwise, people would still be able to do:
```fsharp
Program.mkSimple ...
|> Program.withErrorHandler someHandler
```
and Fabulous won't have access to the setted error handler.

Hiding Elmish.Program means I will need to rewrite every functions `mkSimple`, `mkProgram`, `withErrorHandler`, `with...`

And without that, Elmish loses a lot of its attractions.

**Comment by MangelMaxime** [comment]
Just for info, I linked the issue for the record.

Because if this is not the first time it happens perhaps there is something to be done. And the case of  Fabulous, is probably not the only one. 

I suspect that if someone else is working on a new "platform" we can encounter the same thing perhaps.

**Comment by et1975** [comment]
The idea is to unify on the logging mechanism, which as you all know is highly platform-specific, but not in Elmish, in your own code and make Elmish use it *in addition to your own code*, not the other way around. [Elmish logging](https://github.com/elmish/elmish/blob/master/src/prelude.fs#L9) is absolutely nothing special to get attached to.

> And if I reimplement this logging mecanism, I will also need to completely hide the Elmish.Program module.

I don't see how one follows from the other.

**Comment by TimLariviere** [comment]
I'm not looking to use the Elmish logging in particular, I just want to plug my own error reports inside the same channel provided by Elmish with `program.onError`.

(Because reporting both Elmish outputs and Fabulous outputs in the same channel is what makes more sense in my case)

It's up to the user to choose where these reports will be logged.

>The idea is to unify on the logging mechanism, which as you all know is highly platform-specific, but not in Elmish, in your own code and make Elmish use it in addition to your own code, not the other way around

If I want to unify the logging mechanism at the Fabulous level, I need to prevent my users from accessing the `Elmish.Program` module.
Because otherwise, they can set their error handler for Elmish but not for Fabulous.

And today, writing
```fsharp
Program.mkProgram ...
|> Program.withErrorHandler errorHandler
|> Program.run
```
is widespread and a lot of documentation refers to it.

This means I have 2 options to ensure that my users always goes through Fabulous instead of being able to go through Elmish :
- Provide my own Program module with all the same functions (`mkSimple`, `mkProgram`, `withErrorHandler`, `with...`) and tell them to not use `open Elmish` (highly risky and duplicating most of Elmish.Program for nothing)
- Force them to always give me an error handler when calling `run`

E.g.
```fsharp
Elmish.Program.mkProgram ...
|> Elmish.Program.withErrorHandler errorHandler // This one becomes useless because Fabulous will replace it. You better know you shouldn't do that, the compiler won't help you
|> FabulousProgram.run errorHandler
```

Both options have downsides, where just a simple access to `onError` (for library authors) would be perfect.
```fsharp
Program.mkProgram ...
|> Program.withErrorHandler errorHandler
|> FabulousProgram.run
```

**Comment by dbrattli** [comment]
Yes, same problem with Elmish.Streams. It's a real pain for application developer having to stitch logging into Elmish and having to do the same with every Elmish extension they use. Closing the issues do not make the problem go away 😉

**Comment by et1975** [comment]
```
Program.mkProgram ...
|> Program.withErrorHandler errorHandler
|> FabulousProgram.run
```

Hmm, but it does not prevent you from mapping the handler again inside your `run` function, does it?

**Comment by TimLariviere** [comment]
I can replace the `errorHandler` with my own, but that's not what I want.
I want to access the one provided by the user.

And with that code, Elmish is aware of the user's `errorHandler`, Fabulous is not.

**Comment by et1975** [comment]
ok, let's add this.

**Comment by et1975** [comment]
@dbrattli 
> Closing the issues do not make the problem go away wink

Do you want to get banned? Cause that's how you get banned ;)

**Comment by et1975** [comment]
Since this is a non-breaking change we can do a minor release, anyone wants to do a PR? If not I might get to it on the weekend.

**Comment by et1975** [comment]
Released as 3.0.4.

**Comment by TimLariviere** [comment]
@et1975 Thank you!
